src/api/app/controllers/reports_controller.rb

```diff
@@ -0,0 +1,92 @@
+class ReportsController < ApplicationController
+  validate_action show: { response: :report }
+  validate_action create: { method: :post, request: :report }
+  validate_action update: { method: :put, request: :report }
+
+  before_action :set_report, only: %i[show update destroy]
+
+  # GET /reports
+  def index
+    @reports = Report.order(:id)
+    authorize @reports
+
+    user = User.session
+    if user.present? && !(user.admin? || user.moderator? || user.staff?)
+      @reports = @reports.where(reporter: user)
+    end
+
+    filter_reports
+
+    @reports = Kaminari.paginate_array(@reports, offset: params.fetch(:offset, 0).to_i, limit: params.fetch(:limit, 25).to_i)
```

Comment:
You first fetch *all* records into memory to then paginate the "array". Why not fetch only the records matching your filters?